pandas/core/generic.py

```diff
@@ -6827,6 +6827,14 @@ def interpolate(self, method='linear', axis=0, limit=None, inplace=False,
             index = np.arange(len(_maybe_transposed_self._get_axis(alt_ax)))
         else:
             index = _maybe_transposed_self._get_axis(alt_ax)
+            methods = {"index", "values", "nearest", "time"}
+            num_dt_flg = is_numeric_dtype(index) or is_datetime64_dtype(index)
```

Comment:
actually timedelta and maybe period
are acceptable here (though i doubt we have tests for these); can u see

Reply:
So I tested this and created a summary. I tried all the interpolation methods for the following indexes

1. `Int64Index` 
2. `Index`(by supplying an object list)
3. `DatetimeIndex` without timezone
4. `DatetimeIndex` with timezone
5. `PeriodIndex` (created using `period_range`)
6. `IntervalIndex` (created using `interval_range`)
7. `TimedeltaIndex` (created using `timedelta_range`)

My input data is -

```
df = pd.DataFrame({'a': [1, 2, pd.np.nan, 3]})
```

The results are as follows-

1. `Int64Index` 

- `method=time` failed, which is expected
- `method=cubic` failed, which I don't understand completely why, but I think it's coming from scipy's end
- all other methods ran successfully

2. `Index`(by supplying an object list)

- except `method=linear`, all others failed, which is expected and was the original issue filed
- Common errors were

```python
time - ValueError: time-weighted interpolation only works on Series or DataFrames with a DatetimeIndex
index, values - TypeError: Cannot cast array data from dtype('O') to dtype('float64') according to the rule 'safe'
nearest, piecewise_polynomial, from_derivatives, akima - TypeError: unsupported operand type(s) for -: 'str' and 'str'
zero, slinear - ValueError: could not convert string to float: 'a'
quadratic, cubic, barycentric, krogh, polynomial, spline - TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
pchip - ValueError: object arrays are not supported
```

3. `DatetimeIndex` without timezone

- `method=cubic` failed, which I don't understand completely why, but I think it's coming from scipy's end. This is same as 1st point.
- All other methods ran successfully.

4. `DatetimeIndex` with timezone

- Same as 3rd point

5. `PeriodIndex` (created using `period_range`)

- Same as 2nd point
- Common errors were

```python
time, index, values - TypeError: Cannot cast array data from dtype('O') to dtype('float64') according to the rule 'safe'
nearest - TypeError: unsupported operand type(s) for /: 'Period' and 'float'
piecewise_polynomial, from_derivatives, akima - TypeError: Invalid comparison between Day and int
zero, slinear - TypeError: float() argument must be a string or a number, not 'Period'
quadratic, cubic, barycentric, polynomial, spline - TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
krogh - TypeError: unsupported operand type(s) for /: 'float' and 'Day'
pchip - ValueError: object arrays are not supported
```

6. `IntervalIndex` (created using `interval_range`)

- Same as 2nd point
- Got the following errors

```python
time - ValueError: time-weighted interpolation only works on Series or DataFrames with a DatetimeIndex
index, values - TypeError: Cannot cast array data from dtype('O') to dtype('float64') according to the rule 'safe'
nearest, piecewise_polynomial, from_derivatives, akima - TypeError: unsupported operand type(s) for +: 'pandas._libs.interval.Interval' and pandas._libs.interval.Interval'
zero, slinear - TypeError: float() argument must be a string or a number, not 'pandas._libs.interval.Interval'
quadratic, cubic, barycentric, polynomial, spline - TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
krogh - TTypeError: unsupported operand type(s) for -: 'pandas._libs.interval.Interval' and 'pandas._libs.interval.Interval'
pchip - ValueError: object arrays are not supported
```

7. `TimedeltaIndex` (created using `timedelta_range`)

- `linear`, `time`, `index`, `values` and `pchip` worked successfully
- all other failed with the following errors-

```python
nearest - TypeError: Cannot cast array data from dtype('float64') to dtype('<m8[ns]') according to the rule 'safe'
piecewise_polynomial, from_derivatives, akima - TypeError: Cannot cast ufunc less_equal input from dtype('<m8[ns]') to dtype('<m8') with casting rule 'same_kind'
zero, slinear, #TypeError: invalid type promotion
quadratic, cubic, barycentric, polynomial, spline - TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
krogh - TypeError: ufunc true_divide cannot use operands with types dtype('float64') and dtype('<m8[ns]')
```

-----

- To me it seemed like there was no consistency in the errors but I wasn't sure if it was intentional.
- It doesn't look like all of the cases are being handled in the code and tests.
- I wasn't sure on which indexes these methods are supposed to work. Saw a test where datetime index was being used with a cubic method so just doing a ```is_numeric_dtype``` check was failing those tests. So added ```is_datetime64_dtype``` as well. Did the above analysis to check where all these 2 dtypes are working properly.

Sorry for this long summary. Let me know what changes I have to make.

Reply:
so i guess we don't support timedelta / period indices currently. pls create a test with assert that these fail (ideally the failure message will be consistent and rationale; this may not be easily fixable here, but adding the tests would be great - you can always xfail them).

i guess the same for other types as well.